### PR TITLE
Add CloseHandle for IPC server

### PR DESCRIPTION
**Pull request by pinkisemils**
Hiya, as we're moving to use an `IPC` from websockets, #148 struck once more. These changes address that.

### CODE REVIEW

File: ipc/src/server.rs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 use std;
-use std::sync::Arc;
+use std::sync::{Arc, Mutex};
```
Comment (tomusdrw) [inline ipc/src/server.rs]: Could you use `parking_lot::Mutex` instead? we use it all over the codebase.

File: ipc/src/server.rs [hunk 2]
```diff
@@ -194,13 +195,25 @@ impl<M: Metadata, S: Middleware<M>> ServerBuilder<M, S> {
...
-					.map(|_| ())
+					.map(|_| {
+						let _ = wait_signal.send(());
+						()
+					})
 					.map_err(|_| ())
 			)
 		});
 
+		let handle = InnerHandles{
```
Comment (tomusdrw) [inline ipc/src/server.rs]: space before `{` missing

File: ipc/src/server.rs [hunk 3]
```diff
@@ -209,35 +222,63 @@ impl<M: Metadata, S: Middleware<M>> ServerBuilder<M, S> {
...
-		self.stop.take().map(|stop| stop.send(()));
-		self.remote.take().unwrap().close();
-		self.clear_file();
+	pub fn close(self) {
+		self.handles.lock().unwrap().close();
+	}
+
+	/// Creates a close handle that can be used to stop the server remotely
+	pub fn close_handle(&self) -> CloseHandle {
+		CloseHandle{
```
Comment (tomusdrw) [inline ipc/src/server.rs]: space missing

File: ipc/src/server.rs [hunk 4]
```diff
@@ -209,35 +222,63 @@ impl<M: Metadata, S: Middleware<M>> ServerBuilder<M, S> {
...
+	pub stop: Option<oneshot::Sender<()>>,
+	path: String,
+}
+
+impl InnerHandles {
+	pub fn close(&mut self) {
+		let _ = self.stop.take().map(|stop| stop.send(()));
+		self.remote.take().map(|remote| remote.close());
+
+		self.stop = None;
```
Comment (tomusdrw) [inline ipc/src/server.rs]: Why is this necessary? Shouldn't `take()` set it to `Null` already?

File: ipc/src/server.rs [hunk 5]
```diff
@@ -209,35 +222,63 @@ impl<M: Metadata, S: Middleware<M>> ServerBuilder<M, S> {
...
+		self.close();
+	}
+}
+/// `CloseHandle` allows one to stop an `IpcServer` remotely.
+pub struct CloseHandle {
+	inner: Arc<Mutex<InnerHandles>>,
+}
+
+impl CloseHandle {
+	/// `close` closes the coressponding `IpcServer` instance.
```
Comment (tomusdrw) [inline ipc/src/server.rs]: Typo in `corresponding`

File: ipc/src/server.rs [hunk 6]
```diff
@@ -209,35 +222,63 @@ impl<M: Metadata, S: Middleware<M>> ServerBuilder<M, S> {
...
+		self.wait_handle.take().map(|wait_receiver| wait_receiver.wait());
 	}
 
-	/// Remove socket file
-	fn clear_file(&self) {
+}
+
+
+struct InnerHandles {
+	pub remote: Option<reactor::Remote>,
```
Comment (tomusdrw) [inline ipc/src/server.rs]: It doesn't have to be `pub`, it's still accessible anyway.

**Comment by pinkisemils** [comment]
Hopefully, the minor grumbles should be fixed now.
